client/js/core/protocol.js

```diff
@@ -1,43 +1,269 @@
-		threeVecFromProto(proto.HalfExtents),
-		threeVecFromProto(proto.CenterOffset)
-	)
+Protocol.parseVec3Float = function(result) {
+	return new THREE.Vector3(result.x, result.y, result.z);
 };
 
-Protocol.parseEntityState = function(proto) {
-	return new EntityState(proto.EntityId, Protocol.parseBody(proto.Body), proto.LastUpdated);
+Protocol.parseHealth = function(result) {
+	new Health(result.life)
 };
 
-Protocol.parseHealth = function(proto) {
-	return new Health(proto.Life);
+Protocol.unmarshalMessage = function(offset, dataView) {
+	var fields = PROTO.messages[dataView.getUint8(offset)];
+	offset++;
+	var type = Protocol.unmarshalFields(offset, dataView, fields);
+	return { value: type.value, read: type.read + 1 };
 };
 
-Protocol.parseBioticState = function(proto) {
-	return new BioticState(Protocol.parseEntityState(proto.EntityState),
-			Protocol.parseHealth(proto.Health));
+Protocol.unmarshalFields = function(offset, dataView, fields) {
+	var result = {};
+	var totalRead = 0;
+	var fieldData;
+	for (var i = 0; i < fields.length; i++) {
+		fieldData = Protocol.unmarshalField(offset, dataView, result, fields[i]);
+		result[fields[i].name] = fieldData.value;
+		offset += fieldData.read;
+		totalRead += fieldData.read;
+	}
+
+	return { value: result, read: totalRead };
 };
 
-Protocol.parseWorldItemState = function(proto) {
-	return new WorldItemState(Protocol.parseEntityState(proto.EntityState), proto.ItemKind);
+Protocol.unmarshalField = function(offset, dataView, result, field) {
+	function retVal(val, off) {
+		return { value: val, read: off };
+	}
+
+	var inner;
+	switch (field.type) {
+		case 0:
+			throw "I don't know how to unmarshal this unknown field";
+		case 1: // byte
+			return retVal(dataView.getUint8(offset), 1);
+		case 2: // slice
+			return Protocol.unmarshalByteArray(offset, dataView);
+		case 3: // int
+			return Protocol.unmarshalInt(offset, dataView);
+		case 4: // float
+			return Protocol.unmarshalFloat64(offset, dataView);
+		case 5: // string
+			return Protocol.unmarshalString(offset, dataView);
+		case 6: // bool
+			return retVal(dataView.getUint8(offset) === 1, 1);
+		case 7: // interface
+			var struct = Protocol.unmarshalFields(offset, dataView, PROTO.kinds[result.kind]);
+			result.state = struct.value;
+			Protocol.parseState(result);
+			return retVal(result.state, struct.read + 1);
+		case 8: // struct
+			inner = Protocol.unmarshalFields(offset, dataView, field.fields);
+			return retVal(inner.value, inner.read);
+		case 9: // message
+			return Protocol.unmarshalMessage(offset, dataView);
+	}
 };
 
-function threeVecFromProto(proto) {
-	return new THREE.Vector3(proto.X || 0, proto.Y || 0, proto.Z || 0);
-}
+Protocol.marshalMessage = function(message, data) {
+	var fields = PROTO.messages[message];
+	var buf = new ArrayBuffer(1);
+	var dataView = new DataView(buf);
+	dataView.setUint8(0, message);
+	if (fields.length > 0) {
+		buf = Protocol.append(buf, Protocol.marshalFields(fields, data, 0).buf);
```

Comment:
Should probably like assert (fields.length == data.length) or something. I can see myself doing that by accident.